### ISSUE TITLE
Meta Learning Shared Hierarchies

### ISSUE THREAD

**Issue by Capybara-monkey** [natsu-summer72/RL_survays#1]
## 概要
Shared primitive(根本にある何か?)を利用することで，未知のタスクに対するサンプル効率を改善するメタラーニング手法。階層的な方策を学習。

### 解決したい課題
人間は，新しいタスクに直面したときに過去の経験を有効利用できるのに対し，
強化学習では，様々なタスクをそれぞれ別の問題として扱ってしまい，効率が悪い。
  → 別のタスクで学習したことを，別のタスクの学習に活かすことはできないか？

### Contributions
複数のサブ方策を共有し，それらをマスター方策によって切り替えるモデルを提案(MLSH)。

### 提案手法(概要)
* 複数のサブ方策をパラメータ Φ で表現し，マスター方策を θ で表現する。
* パラメータΦは全てのタスクで共有するパラメータで，テスト時には固定。

#### 図1
<img width="245" alt="structure" src="https://user-images.githubusercontent.com/47890921/56749650-19bff080-67bd-11e9-9e09-2b86d10c48be.png">

* タスクの選択確率 P_M における，T ステップの収益の期待値を最大化する。

#### 目的関数
![objective](https://latex.codecogs.com/gif.latex?{\rm&space;maximize}_\phi&space;E_{M&space;\sim&space;P_M&space;,t=0...T-1}[R])

つまり，様々なタスクにおける収益を最大化するようなパラメータ Φ を事前学習により獲得し，
テストの際の新しいタスクに対しては，パラメータ θ を調整することで適応する。

Φはベクトル， Φ_1, Φ_2 もベクトル(それぞれが，別の方策を表す)
![phi_vector](https://latex.codecogs.com/gif.latex?\bm{\phi}&space;=&space;[\phi_1,&space;\phi_2,&space;\cdots,&space;\phi_K&space;])

θはΦとは別のNNで，サブ方策を選択(master Policy)

### 提案手法(アルゴリズム)
<img width="433" alt="algo1" src="https://user-images.githubusercontent.com/47890921/56749602-01e86c80-67bd-11e9-9245-717c63bc630b.png">

#### Warmup period
* Φでパラメータ化されたサブ方策を仮定
* ここで収集された経験は，master policyによる報酬であるとみなす。(図2の左)
    → サブ方策の選択は，master policyによる行動であるとみなし，サブ方策により得られる報酬の総和を，master policyの行動による報酬だと考え，パラメータ θ を更新

#### Joint update period
* サブ方策のパラメータ Φ と，master policy のパラメータ θ を両方更新
* Warmup periodと同じように，パラメータ θ を更新するとともに，同じ経験を用いて，サブ方策を最適化する。(ただし，サブ方策は，自身の経験だけを用いて更新される)

#### 図2
<img width="409" alt="structure2" src="https://user-images.githubusercontent.com/47890921/56749874-876c1c80-67bd-11e9-82a2-153f28288be2.png">

#### なぜうまくいくのか？
提案手法では，Warmup period において，Φでパラメータ化されたサブ方策に対して，最適な θ を獲得していると仮定している。よって，最終的に得られた Φ における master policyは， θ を学習しなおしても，同じであるという主張。
重要なのは，Φ は θが最適に近い場合のみで更新されるという点である。
例: 迷路タスクで，2つのゴールがある場合を考える。
* θがランダムな場合
        サブ方策は，2つのゴールの中間に導くような方策となる
* θが最適な場合
        2つのサブ方策が，それぞれ別のゴールへ導くような方策となる

また，サブ方策による方策の選択は N ステップ(固定)続くとし，
報酬は， 1/N 倍して扱う。

#### 実験設定
* マスター方策，サブ方策ともに (64,64) のMLP
* マスター方策は，softmax関数でサブ方策を選択
* それぞれの方策は，PPOにより最適化( learning rate:  0.01 for θ,  0.0003 for Φ )
* batchsize D=2000

#### 実験 1  : 2d Moving Bandits
* 2つのサブ方策が，それぞれ別の行動を選択(図3)
* 1つの方策を学習するよりも，Master方策を学習するほうが早い(図4)
* Warmup periodがない場合は，最初は同じような方策を学習するが，徐々に別の方策へとシフトする。(図4)
* Master方策を副方策と同じタイムスケールで学習すると，，単にshared Policyを学習する場合と同じような結果となる。

#### 図3
![result1](https://user-images.githubusercontent.com/47890921/56779690-0f324500-6817-11e9-9416-dc925b23dc70.png)

#### 図4
![result2](https://user-images.githubusercontent.com/47890921/56779962-5ff66d80-6818-11e9-8220-fde65c623563.png)

### 実験 2 : GridWorld
MLSHを，PPO baseline, Option-Critic(階層型強化学習)と比較

#### 実験環境
* エージェントはある初期位置からスタートするが，ゴール位置はランダムに割り当てられる。

#### 実験設定
* ゴールに到達した場合は +1 の報酬を与える。
* エピソードは，100step続き，Master方策による行動は，25step続く。
#### 実験結果
* Option-CriticはPPOと同程度の報酬しか得られないが，MLSHはそれ以上の報酬を獲得(図4下)
